include/pybind11/common.h

```diff
@@ -396,13 +396,8 @@ template<size_t ...S> struct make_index_sequence_impl <0, S...> { typedef index_
 template<size_t N> using make_index_sequence = typename make_index_sequence_impl<N>::type;
 #endif
 
-#if defined(PYBIND11_CPP17) || defined(_MSC_VER)
-using std::bool_constant;
-using std::negation;
-#else
 template <bool B> using bool_constant = std::integral_constant<bool, B>;
 template <class T> using negation = bool_constant<!T::value>;
```

Comment:
Replace this line with the following to fix the MSVC internal compiler error:
```C++
template <typename T> struct negation : bool_constant<!T::value> { };
```

The other compilers can bet away with just the alias, but I don't think it's worth splitting up the implementation over it.